### PR TITLE
Ml vs collab

### PR DESCRIPTION
**Pull request by jeronstone**

### CODE REVIEW

File: components/communications/esp_server.py [hunk 1]
```diff
@@ -125,16 +126,26 @@ def message_received(client, server: WebsocketServer, message):
         client_server.send_print_message(client['teamName'],
                                          get_mission_message(client['teamType'], message['type'], message['message']))
     if message['op'] == 'prediction_request':
+        
+        # TODO TESTING VAR
+        use_jetson = False
```
Comment (ForrestFire0) [inline components/communications/esp_server.py:130]: lol

File: components/machinelearning/ml.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                if cap.isOpened():
+                    ret, frame = cap.read()
+                else:
+                    raise Exception("Could not get image from WiFiCam (cv2)")
+                
+                logging.debug('Entering preprocess...' )
+                picture = preprocess(frame)
+                results = self.handler(picture, team_name, model_index)
+            except Exception as e:
+                logging.debug('ML FAILED :(')
```
Comment (ForrestFire0) [inline components/machinelearning/ml.py:97]: can we add a client message when there is an error? With the error

File: installdependencies.sh
Comment (ForrestFire0) [inline installdependencies.sh:1]: ???

**Review by ForrestFire0 — APPROVED** [review]
Looks great! Lets add an error message when the hander fails with an error so we might have a chance at knowing what is going on just from the VS output.

Nice work Josh!

File: components/machinelearning/ml.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    def handler(self, image, team_name, model_index):
+        model_fi = None
+        for entry in os.scandir(self.model_dir):
+            if entry.name.startswith(team_name) and int(entry.name.split('_')[1]) == model_index:
+                model_fi = entry.name
+                break
+
+        if model_fi is None:
+            raise Exception(f"Cound not find model for team: {team_name} with model index: {model_index}; Available models: {', '.join([entry.name for entry in os.scandir(self.model_dir)])}")
```
Comment (anonymous563) [inline components/machinelearning/ml.py]: typo: Could, not Cound

**Pull request merged**